File: selection/_022/Main.java
package selection._022;

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        double p = sc.nextDouble();
        sc.close();

        double l = -1, r = 1000000000000000000.0;
        while (r - l > 0.0000000001) {
            double mid = (l + r) / 2;
            if (primeFunction(mid, p) >= 0.0) {
                r = mid;
            } else {
                l = mid;
            }
        }

        if (r >= 0.0) {
            System.out.println(function(r, p));
        } else {
            System.out.println(p);
        }
    }

    static double primeFunction(double x, double p) {
        return 1 + p * Math.pow(2, -x / 1.5) * Math.log(Math.pow(2, -2.0 / 3.0));
    }

    static double function(double x, double p) {
        return x + p * Math.pow(2, -x / 1.5);
    }
}

File: selection/_022/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var p float64
	fmt.Scan(&p)

	l, r := 0.0, 1e18
	for i := 0; i < 1000; i++ {
		ml, mr := (2*l+r)/3, (l+2*r)/3
		if f(ml, p) <= f(mr, p) {
			r = mr
		} else {
			l = ml
		}
	}

	fmt.Println(f(l, p))
}

func f(x, p float64) float64 {
	return x + p/(math.Pow(2, x/1.5))
}
